proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,629 @@
+
+* Use ``''`` to delimit multiline strings, which has the benefit of being a parse error without ``MultilineStrings``
+
+* Enable any number of ``"""+`` quotes to delimit multiline strings
+
+* Reuse single-quoted ``"`` for multiline syntax
+
+  * Would require escaping double quotes in the multiline string, which, while not a major part of the proposal, is a nice bonus
+
+* New ``[s|foo|]`` construct that embeds a multiline literal string with no TH (so *not* a quasiquoter, but reusing the same syntax)
+
+* Support arbitrary terminators, like Bash's heredocs
+
+  ::
+
+    x = <<EOF
+    line 1
+    line 2
+    line 3
+    EOF
+
+  * Everyone will use a different terminator, which I think would contribute to a reduction in overall readability
+  * I think ``"""`` is an uncommon enough delimiter, and it can be escaped, that I don't think this is necessary
+
+* Use some delimiter to start the string, but use layout indentation rules to dictate when the string ends
+
+* Strip trailing whitespace in post-processing
+
+  * Nice to have, but not necessary. I think it would be better to keep post-processing as minimal as possible, and it doesn't seem as common as removing leading whitespace.
+
+Out of scope
+~~~~~~~~~~~~
+
+* String interpolation
+  * See https://github.com/ghc-proposals/ghc-proposals/pull/570
+  * One way this proposal can work with raw strings is by allowing both ``s"..."`` and ``s"""..."""`` syntaxes. In general, any raw strings proposal that works with the current double quoted string syntax should be able to work with a triple-quoted string syntax as well, since the proposed triple-quoted string syntax desugars to a single-quoted string.
+
+* "Raw" strings (without escaping)
+  * To an extent, this proposal already helps this a little bit, since double quotes no longer need to be escaped within a triple-quoted string. But this proposal doesn't address needing to escape backslashes.
+  * This is particularly useful for regexes or any other situation where the backslash character is useful as an actual character.
+  * One way this proposal can work with raw strings is by allowing both ``r"..."`` and ``r"""..."""`` syntaxes. See comment in "String interpolation".
+
+Comparisons with other languages
+--------------------------------
+
+* Java
+
+  * As mentioned in the beginning, this proposal draws a lot from Java.
+
+  * Java strips trailing whitespace. See "Strip trailing whitespace" under "Alternatives".
+
+  * Java defines the content to start after the first newline after the opening ``"""``, and disallows any non-whitespace characters after the opening delimiter. Instead of adding this restriction, we added the rule to remove exactly one newline from the beginning of the string, if one exists. This allows people to start the multiline string on the same line, enabling one-line strings to use the syntax, e.g. ``"""A string using "unescaped" quotes"""``.
```

Comment:
> allows people to start the multiline string on the same line, enabling one-line strings to use the syntax

Aha, thank you, this helps explain the subtle rules around discarding leading whitespace. Have you considered just splitting the behavior on whether the `"""` string spans one or many lines?

* a single-line `"""` string is just like a regular string, but may contain the literal `"` character
* a multi-line `"""` string must not have any non-whitespace characters after the opening delimiter (like java). And we would probably strip leading and trailing newlines.

Reply:
Have you seen my comment in the thread above? https://github.com/ghc-proposals/ghc-proposals/pull/569#discussion_r1343499643

Your suggestion would prevent users from doing the last snippet. What would be the reason to change this? It seems like you're just removing one rule (remove exactly one newline at the beginning) and adding two more rules.

Reply:
Yes, it would prevent people from writing your last example.

```
"""asdf
   asdf
   asdf"""
```

The whole point of this proposal is to give people a more WISYWIG syntax for creating multiline strings (h/t @maxigit). In the above example I think it's fairly clear what the resulting string should be, but take your other example from that comment

```
s = """ foo
    bar
    """
```

There is no obvious answer to how that string literal should be parsed, and the actual result (`" foo\nbar"`) does not match the visual layout at all. It gets worse if you actually try to align the lines, e.g.

```
s = """ foo
        bar
    """
```

Visually, this looks like it should be parsed as either `" foo\n bar"` or `"foo\nbar"` (probably the former), but actually your proposal would parse it as `" foo\nbar"`. Note that, no matter how you indent the second line, the string will be parsed the same. This is weird!!

Best to just rule out this set of corner cases entirely, and ask people to write the more regular

```
"""
asdf
asdf
asdf
"""
```

for your first example.

Reply:
@ChickenProp You had originally suggested allowing characters on the same line as the opening `"""` delimiter. What are your thoughts on the example above (copied here):
```
s = """ foo
        bar
    """
```
I think the commenter makes a good point that it might be a bit surprising for this to resolve to `" foo\nbar"`

Reply:
I can see it being a bit surprising, and something like "leading whitespace on the first line is stripped but not used for common prefix identification" could be reasonable. (As always, if you want it, you can open with `\&`.) I don't feel too strongly about that; but I do want to be able to put text on the opening (and closing) lines.

(I agree it's possible to write things that are unclear, if you can do that. Personally I don't think this is a case where it's worth removing something useful because someone could do silly things with it.)